src/main/scala/is/hail/keytable/KeyTable.scala

```diff
@@ -62,43 +36,45 @@ object KeyTable {
   }
 }
 
-case class KeyTable(@transient hc: HailContext, @transient rdd: RDD[(Annotation, Annotation)],
-  keySignature: TStruct, valueSignature: TStruct) extends Serializable {
-  require(fieldNames.areDistinct())
+case class KeyTable(hc: HailContext, rdd: RDD[Annotation],
+  signature: TStruct, keyNames: Array[String]) {
 
-  private val localValueSignature = valueSignature
-  val (signature, mergeKeyAndValue) = keySignature.merge(localValueSignature)
+  if (!fieldNames.areDistinct())
+    fatal(s"Column names are not distinct: ${fieldNames.duplicates().mkString(", ")}")
+  if (!keyNames.areDistinct())
+    fatal(s"Key names are not distinct: ${keyNames.duplicates().mkString(", ")}")
+  if (!keyNames.forall(fieldNames.contains(_)))
+    fatal(s"Key names found that are not column names: ${keyNames.filterNot(fieldNames.contains(_)).mkString(", ")}")
 
   def fields = signature.fields
 
-  def keySchema = keySignature.schema
-
-  def valueSchema = valueSignature.schema
+  def keyFields = keyNames.map(signature.fieldIdx).map(i => fields(i))
 
   def schema = signature.schema
 
-  def keyNames = keySignature.fields.map(_.name).toArray
-
-  def valueNames = valueSignature.fields.map(_.name).toArray
-
-  def fieldNames = keyNames ++ valueNames
+  def fieldNames = fields.map(_.name).toArray
 
   def nRows = rdd.count()
 
   def nFields = fields.length
 
-  def nKeys = keySignature.size
-
-  def nValues = valueSignature.size
+  def nKeys = keyNames.length
 
   def typeCheck() = {
-    rdd.forall { case (k, v) => keySignature.typeCheck(k) && valueSignature.typeCheck(v) } &&
-      KeyTable.toSingleRDD(rdd, nKeys, nValues).forall { a =>
-        signature.typeCheck(a)
-      }
+    val localSignature = signature
+    rdd.forall { a => localSignature.typeCheck(a) }
   }
 
-  def printSchema(): Unit = println(signature.toPrettyString())
+  def withKeys(): RDD[(Annotation, Annotation)] = {
+    val keyIndices = keyFields.map(_.index)
+    val nFieldsLocal = nFields
+
+    rdd.map { a =>
+      val r = KeyTable.annotationToSeq(a, nFieldsLocal)
```

Comment:
When would the top-level row of a KeyTable be null?  I'm wondering if this is necessary.